crates/bevy_ecs/src/query/state.rs

```diff
@@ -438,24 +438,23 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
         last_change_tick: u32,
         change_tick: u32,
     ) -> Result<[ROQueryItem<'w, Q>; N], QueryEntityError> {
-        // SAFETY: fetch is read-only
-        // and world must be validated
-        let array_of_results = entities.map(|entity| {
-            self.as_readonly()
-                .get_unchecked_manual(world, entity, last_change_tick, change_tick)
-        });
+        let mut values = [(); N].map(|_| MaybeUninit::uninit());
 
-        // TODO: Replace with TryMap once https://github.com/rust-lang/rust/issues/79711 is stabilized
-        // If any of the get calls failed, bubble up the error
-        for result in &array_of_results {
-            match result {
-                Ok(_) => (),
-                Err(error) => return Err(*error),
-            }
+        for (i, entity) in entities.into_iter().enumerate() {
+            // SAFETY: fetch is read-only
+            // and world must be validated
+            let item = self.as_readonly().get_unchecked_manual(
+                world,
+                entity,
+                last_change_tick,
+                change_tick,
+            )?;
+            values[i] = MaybeUninit::new(item);
```

Comment:
This does a bounds check. It's a bit more verbose, but `slice::get_unchecked_mut` might further speed things up a bit. We know the index must be valid because `i` must be less than `N`.

Reply:
I ended up using an iterator to remove the bounds checks -- I measured the same performance for both iterators and `get_unchecked_mut`.